### ISSUE TITLE
[HOLD for payment 2024-02-20] [$500] iOS - Login - Background image moves downward when an error appears

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#34529]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.25-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Open New Expensify app.
2. Enter an email and tap on "Continue".
3. Enter 6 random numbers (to trigger incorrect magic code error).

## Expected Result:
When the error appears, the background image should remain static.

## Actual Result:
The background image moves downward.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![image](https://github.com/Expensify/App/assets/93399543/ef7d2996-6704-42d1-bf6d-48753872c9da)

https://github.com/Expensify/App/assets/93399543/48a91c18-58b3-467d-b26c-2752ee4eb13b

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011bb5ec1c98737926</li>
        <li>Upwork Job ID: 1746986846306430976</li>
        <li>Last Price Increase: 2024-01-15</li>
<li>Automatic offers: </li>
<ul>
<li>s77rt | Reviewer | 28130620</li>
<li>shahinyan11 | Contributor | 28130621</li>
</ul></li>
    </ul>
</details>

**Comment by Ykumar1415** [comment]
I tried this issue and able to reproduce this issue .

**Comment by shahinyan11** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
iOS - Login - Background image moves downward when an error appears

### What is the root cause of that problem?
The background image has `position: "absolute", bottom: 0` style, and its height is less than the height of the parent container. When the error message appears, the height of the parent container is increased and the image is moved down due to the `bottom: 0` style.
### What changes do you think we should make in order to solve the problem?
We can add  `top` position  which will help us prevent the image from moving down  . To do that we can change the [BackgroundImage](https://github.com/Expensify/App/blob/ee7704626bdd70ed469fb270da65a38680403454/src/pages/signin/SignInPageLayout/BackgroundImage/index.ios.js#L20C1-L31C2) component for ios wit below 
```
function BackgroundImage(props) {
    const styles = useThemeStyles();
    const StyleUtils = useStyleUtils();
    const src = useMemo(() => (props.isSmallScreen ? MobileBackgroundImage : DesktopBackgroundImage), [props.isSmallScreen]);
    const [top, setTop] = useState(null);

    const onLayout = useCallback(({nativeEvent}) => {
        setTop(nativeEvent.layout.y);
    }, []);

    return (
        <Image
            source={src}
            style={[styles.signInBackground, StyleUtils.getWidthStyle(props.width), {top}]}
            onLayout={onLayout}
        />
    );
}
```

### What alternative solutions did you explore? (Optional)
Since the height of the image is given by a static value ( 700 ), we can also calculate the top position using the following combination
```
const {windowHeight} = useWindowDimensions();
const insets = useSafeAreaInsets()
const top = windowHeight - insets.top - 700 
```

**Comment by pipolodev** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
iOS - Login - Background image moves downward when an error appears

### What changes do you think we should make in order to solve the problem?
The styles in [SignInPageContent()](https://github.com/Expensify/App/blob/main/src/pages/signin/SignInPageLayout/SignInPageContent.js) include excessive `mb` and `mt` properties, causing the content to extend beyond the parent container and pushing the image downward. To resolve this issue, I removed one `mb` when the screen its small to create additional space.

Original Line [here](https://github.com/Expensify/App/blame/f7e815f2d36253496f6f85cf8a36cccfc3ce52a2/src/pages/signin/SignInPageLayout/SignInPageContent.js#L51C1-L51C62).

Replacing that line whit this:
`<View style={[styles.flexGrow2, !isSmallScreenWidth ? styles.mb8 : null]}>`

 And effectively resolving the problem.

### What alternative solutions did you explore? (Optional)
N/A

**Comment by s77rt** [comment]
@tienifr Thanks for the proposal. Your RCA makes sense. As for the solution can we find something more permanent? i.e. won't break if we add more text below the error message.

**Comment by s77rt** [comment]
@shahinyan11 Thanks for the proposal. Your RCA is correct. Regarding the solution won't using `top` instead of `bottom` just create a different kind of problems, i.e. when an error message appears an empty space will be created (when the parent height exceeds the background height)

**Comment by s77rt** [comment]
@pipolodev Thanks for the proposal. Unfortunately I don't your RCA is correct.

**Comment by MichaelBuhler** [comment]
Contributor details
Your Expensify account email: michael@michaelbuhler.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01888d8c892c42b3b1

**Comment by MichaelBuhler** [comment]
Contributor details
Your Expensify account email: [michael@michaelbuhler.com](mailto:michael@michaelbuhler.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~01888d8c892c42b3b1

**Comment by MichaelBuhler** [comment]
Contributor details
Your Expensify account email: michael@michaelbuhler.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01888d8c892c42b3b1

**Comment by shahinyan11** [comment]
> when an error message appears an empty space will be created

@s77rt  Sorry, but I didn't understand where the empty space will be created ? I checked, there is no empty space.

**Comment by s77rt** [comment]
@tienifr I think that makes sense. Can you please check if that value is enough for Android as well?

**Comment by s77rt** [comment]
@shahinyan11 If we pin the image to the bottom increasing the parent height will create space at the top. Similarly if we pin the image at the top increasing the parent height will create space at the bottom. It's safer to pin the image to the bottom as the extra is space is seen as the "sky" where if we pin the background to the top it will feel much buggy.

Consider this example where I added 400px just for testing:

| Using bottom | Using top |
|--------------|-----------|
|   <img width="572" alt="Screenshot 2024-01-18 at 12 53 51 AM" src="https://github.com/Expensify/App/assets/16493223/6981de6d-2991-4752-ade9-038fb31beae2">           |     <img width="572" alt="Screenshot 2024-01-18 at 12 41 19 AM" src="https://github.com/Expensify/App/assets/16493223/e9f1f6d4-f0c8-4043-b504-ca4a685d3489">      |

**Comment by shahinyan11** [comment]
> If we pin the image to the bottom increasing the parent height will create space at the top. Similarly if we pin the image at the top increasing the parent height will create space at the bottom. It's safer to pin the image to the bottom as the extra space is seen as the "sky" where if we pin the background to the top it will feel much buggy.

@s77rt  We should not set only `bottom` or only `top` . We should set both, then the space will not appear . 
Maybe you removed the `bottom: 0` from styles. there is no need to do this
Are you tested by just changing my provided code ? Did you check by simply changing the code I provided ?

**Comment by shahinyan11** [comment]
> @tienifr I think that makes sense. Can you please check if that value is enough for Android as well?

@s77rt  here is the @tienifr's solution result on IOS 

| signInContentMinHeight: 800 | signInContentMinHeight: 850 |
|--------|--------|
| ![Screenshot 2024-01-18 at 12 37 40](https://github.com/Expensify/App/assets/46921547/b7456592-0038-4adb-b7e6-f4cde909cec8) | ![Screenshot 2024-01-18 at 12 37 28](https://github.com/Expensify/App/assets/46921547/c67164bd-dc4f-4998-a1f0-f04aee1a8d00) |

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/34529#issuecomment-1894169543). I have added new solution and moved the old one as alternatives

**Comment by s77rt** [comment]
@shahinyan11

> We should not set only bottom or only top . We should set both

Setting both will cause the image to stretch 

> And remove bottom and set only top for bg image

What value should we set for `top`?

> Or we can increase bg image minHeight to 750

I think this will also cause the image to stretch, the buildings will appear a little bigger

**Comment by shahinyan11** [comment]
@MichaelBuhler I have added a videoы in updated proposal where you can see the result of both. Have you checked  ?

**Comment by shahinyan11** [comment]
@MichaelBuhler I have added a videos in updated proposal where you can see the result of both. Have you checked  ?
What value should we set for top?

> What value should we set for `top`?

If we do not want set any static value for `top` we can kip `bottom:0`  in styles then In `onLayout` event we can calculate   current top position and add following style `{bottom:0, top: CALCULATED VALUE}`

**Comment by s77rt** [comment]
I agree with @tienifr here. This is a minor bug that should be fixed a very simple solution.

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/34529#issuecomment-1893281868)

**Comment by shahinyan11** [comment]
@s77rt @tienifr Sorry, but I tested the solution by setting SignInContentMinHeight to 850 or 900, but that didn't solve the problem. Maybe I did something wrong . Could you please provide an example with this solution ?

**Comment by shahinyan11** [comment]
@s77rt @tienifr @rlinoz  
Disadvantages of the chosen proposal by C+

1. The 900 is minimal value of `signInContentMinHeight` which will work for both screens `email or phone` and `enter magic code` and it seriously affects to design
2. We have an error message that is long and even `signInContentMinHeight:900` doesn't work in this case. Below you can see the example

https://github.com/Expensify/App/assets/46921547/8ea65fc1-3aa7-43c6-a2cf-1375177627b3

**Comment by s77rt** [comment]
@shahinyan11 Thanks for mentioning this case. I think if we are going to fix this issue we should fix all the cases or just do nothing. Yet I still don't think using top is best here since it can cause this issue https://github.com/Expensify/App/issues/34529#issuecomment-1897510602

**Comment by s77rt** [comment]
@tienifr Let's see if we can fix all the cases here

**Comment by s77rt** [comment]
@rlinoz False positive 😅 No proposal to review here, still working on it...

**Comment by shahinyan11** [comment]
> Yet I still don't think using top is best here since it can cause this issue [#34529 (comment)](https://github.com/Expensify/App/issues/34529#issuecomment-1897510602)

@s77rt But I have provided examples where all works correct without any issues. Did you tested two ways from my updated proposal

**Comment by s77rt** [comment]
@tienifr 


https://github.com/Expensify/App/assets/16493223/cff73146-f7f7-47a8-ae93-5e4d1190febd

**Comment by shahinyan11** [comment]
> Option 1

@s77rt  I will create new PR. I think you did something wrong for `option 1`. The new View element should have styles `bottom:0` and `height: "50%"` ?

**Comment by s77rt** [comment]
@shahinyan11 Why height: 50?

**Comment by shahinyan11** [comment]
> @shahinyan11 Why height: 50?

@s77rt so that it is not visible at the top of the background image, but at the same time fills the entire bottom part for any height of container

**Comment by s77rt** [comment]
@shahinyan11 I think this is feeling workaround-ish, there is nothing special about 50% besides being a random value that just works.

**Comment by s77rt** [comment]
@tienifr I actually missed step 2 I thought only step 1 is required for iOS. Removing the bottom style makes the city "flying". The building should be on the bottom and the remaining space should be used for the "sky" not the "ground".

**Comment by shahinyan11** [comment]
> @shahinyan11 I think this is feeling workaround-ish, there is nothing special about 50% besides being a random value that just works.

@s77rt Sorry, but your answer is a little strange for me. I already explained why 50%. I can't understand what exactly you don't like.  Wasn't 850 or 900 random for the minHeight of the container In your previous approved solution ?

**Comment by s77rt** [comment]
Seeing how this is getting complex with low ROI, I asked on slack https://expensify.slack.com/archives/C01GTK53T8Q/p1706048373617079 and I think we should just close this.

cc @rlinoz @zanyrenney

**Comment by rlinoz** [comment]
I agree with closing.

**Comment by shahinyan11** [comment]
@s77rt @rlinoz I think there is a [solution](https://github.com/Expensify/App/issues/34529#issuecomment-1894169543) that fixes the problem and it works well enough to implement it rather than leaving an error. I still don’t understand what are the disadvantages in my proposal, what’s wrong with it?

**Comment by s77rt** [comment]
@tienifr Can you please elaborate on extending the SVG height. How would that look?

**Comment by shahinyan11** [comment]
> @shahinyan11 Using 800px or 850px as min height is self explanatory. Where adding a new view with 50% height is not. It feels a little hacky.

@s77rt We want the image to not stretch or change position, so we can't avoid creating empty space when the height of the container increases.
And my solution just fills that empty space. I don't think this is a hack

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/34529#issuecomment-1894169543).  I updated the solution which to avoid using random `50%`

**Comment by s77rt** [comment]
@shahinyan11 Can you please explain your proposal in plain English without PRs. You should only raise a PR when assigned

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/34529#issuecomment-1894169543). Added explanation
@s77rt Please check

**Comment by s77rt** [comment]
@shahinyan11 I'm seeing unmatched color at the top

<img width="572" alt="Screenshot 2024-01-28 at 5 05 59 PM" src="https://github.com/Expensify/App/assets/16493223/e89fa7e3-17d3-419f-bc7a-36c09f9ec7f5">

**Comment by shahinyan11** [comment]
> @shahinyan11 I'm seeing unmatched color at the top

@s77rt  I'm sure you're missing something. I checked and there is no such thing. It looks like you haven't set the background color for the new container. Did you set `backgroundColor: colors.productDark200`.

**Comment by shahinyan11** [comment]
> All [voted](https://expensify.slack.com/archives/C01GTK53T8Q/p1706048373617079) "Do nothing" except @tienifr

My [solutions](https://github.com/Expensify/App/issues/34529#issuecomment-1894169543) is quite simple and will fix the error completely

**Comment by s77rt** [comment]
@shahinyan11 What's the point of the absolute view? Isn't your solution same as the previous one which is to change the top style of signInBackground to window height - svg height?

**Comment by s77rt** [comment]
@tienifr I don't think we need to change the svg in this case.

**Comment by shahinyan11** [comment]
> @shahinyan11 What's the point of the absolute view? Isn't your solution same as the previous one which is to change the top style of signInBackground to window height - svg height?

1. The new absolute "View" will have the height of the screen and will not be associated with a resizable container. So it will always maintain its height.
2. The image will be in this container and will be placed at the bottom. it helps to keep the existing UI. And the image never shifts down because the height of its container is always the same as the height of the screen.

**Comment by s77rt** [comment]
@shahinyan11 Upon testing I found that the absolute view help with an issue I faced on web with the `top` styling. But why would we need to change the `minHeight` of `signInBackground`? This also results in unwanted change on web

**Comment by shahinyan11** [comment]
> @shahinyan11 Upon testing I found that the absolute view help with an issue I faced on web with the `top` styling. But why would we need to change the `minHeight` of `signInBackground`? This also results in unwanted change on web

If we keep minHeight the image will take full height of container which will change current UI. I think then we need use separate styles for web and ios

**Comment by s77rt** [comment]
@shahinyan11 Can you please attach two images for the diff if we keep `minHeight` vs changing it to `height`

**Comment by s77rt** [comment]
@tienifr Changing the SVG not for the purpose of actually changing it but to fix a bug does not feel right. I'd like to get this fixed with the least footprint possible.

**Comment by shahinyan11** [comment]
> @shahinyan11 Can you please attach two images for the diff if we keep `minHeight` vs changing it to `height`

My bad. I was wrong. If we keep `minHeight` it also works correct and not fill full container

**Comment by shahinyan11** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/34529#issuecomment-1894169543). Updated  step `3`

**Comment by rlinoz** [comment]
Sure, since we found a good solution this is some nice polish, lets do it!

Assigning @shahinyan11

**Comment by shahinyan11** [comment]
@rlinoz My proposal causes regression on the small screens of the Internet. And so I made additional changes to avoid this regression. The latest changes are made to ensure that all changes work only for IOS. Could you please check last changes in [this](https://github.com/Expensify/App/pull/34958/files) PR and approve

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: n/a. This is more like a feature request
- The offending PR has been commented on: n/a
- A discussion in #expensify-bugs has been started: n/a
- Determine if we should create a regression test for this bug: No

**Comment by shahinyan11** [comment]
@zanyrenney Hi. When sending payment do not forget to leave feedback please

**Comment by zanyrenney** [comment]
**payment summary**

@s77rt requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/28130620) (Reviewer) - paid $500
@shahinyan11 requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/28130621) (Contributor) - paid $500
